jsdoc fontification and spellcheck support

lddubeau opened the pull request:
This is the fleshed out version of 

https://github.com/ananthakumaran/typescript.el/tree/experiment/typedoc-tags

I've purposely split it into multiple commits:

1. There was an explicit invocation of the mode hooks in the code. I've removed it because `define-derived-mode` should call it automatically. (It sure does call it for me.) So the explicit invocation was redundant.

2. I've added fontification support for the jsdoc/typedoc tags. It is **very** unlikely to be the final word on this. The fact of the matter is that "jsdoc" is really a family of dialects for writing JavaScript documentation. Typedoc is one member of the family which focuses on the needs of TypeScript. Anyway, there's no definite set of tags for "jsdoc". Such and such version of jsdoc3 supports a definite set of tags, but some people use Google's Closure compiler which supports a different dialect, others run jsdoc2 (they really shouldn't but...) which supports another dialect, Typedoc supports yet another dialect, etc. I've named Elisp symbols so that those which deal with features usually common across dialects have `jsdoc` in their name, and those which are specific to Typedoc have `typedoc` in their name.

3. I've added a custom predicate for `flyspell-prog-mode`. In addition to the default fontification-based logic that `flyspell.el` provides, it excludes from spell-checking strings in `import` statements that refer to module names.

4. I've removed the fontification hack reported in #50.

I see there are CI failures on older versions of Emacs. I don't know yet what the solution for that is going to be.

I've used this code to spell check a fairly large code-base that uses Typedoc as its documentation processing tool.

josteink commented:
Looks like extensive and proper work, as usual. :)

I fully support these changes.

lddubeau commented:
I've added commits to fix the issues we discussed. I've also run spell check on another code base, smaller than the earlier one I used, but I still managed to find a couple issues that I fixed.

One change I forgot to mention in my initial PR: the fontification includes a regexp for tslint flags:

```
// tslint:disable-next-line:blahblah
```

from the word `tslint` onwards this is fontified using `font-lock-preprocessor-face` because to me it is akin to a preprocessor directive... though maybe there are good reasons to want another face for it. Let me know if you have an issue with this.

Once we're in agreement, I plan to rebase the PR here to fold the additional commits I just added into the original 4 commits and merge.

And I just got notice that the tests passed. :tada:

josteink commented on typescript-mode.el:
```diff
@@ -1741,8 +1741,11 @@ and searches for the next token to be highlighted."
+  (if (fboundp 'font-lock-ensure)
+      (font-lock-ensure)
+    (font-lock-fontify-buffer))
   (and
```
This is nice. I like this way better.

It's even documented so at whatever point in the future when we don't feel the need to support Emacs 24.x we can clean it up.

lddubeau replied:
I had a vision of a future me some years from now looking at this code and going "Why the hell did I do this?" So I commented. :)

josteink commented on typescript-mode.el:
```diff
@@ -1756,12 +1756,12 @@ and searches for the next token to be highlighted."
+                                        (nth 8 parse))))
+            (and string-start-pos
+                 (save-match-data
                    ;; Move to back to the start of the string, then past any ws
```
This is absolutely nitpicking, and I in no way mean to imply the tests so far are insufficient.

That said: When you encountered this issue and decided to write a fix for it... Was this a regression that can somehow be verified in a test, and thus prevented in the future?

lddubeau replied:
I noticed the issue when I was doing `M-: (typescript--flyspell-mode-predicate)` to figure out one of the issues I ran into. I noticed that the cursor was moving back, and it was annoying me.

I did not experience a failure while using flyspell through the normal means.

I guess I could still ensure in the test suite that point is not moved by the predicate, which is probably a sensible rule, generally speaking.

josteink commented:
Pristine as always. 

If you can fix up and clean the commits like you said, this is about as perfect as I can see a PR being.

Go do your thing :)
